File: 프로그래머스/2/17687. ［3차］ n진수 게임/［3차］ n진수 게임.cpp
#include <string>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

// i를 n진법으로 바꾸어 리턴 (10이상은 알파벳으로)
string digits = "0123456789ABCDEF";
string change(int i, int n){
    if(i == 0) return "0";
    
    string res = "";
    while(i > 0){
        int r = i % n;
        res += digits[r];
        i /= n;
    }
    
    // 역순으로
    reverse(res.begin(), res.end());
    return res;
}

// 진법 n, 미리 구할 숫자의 갯수 t, 게임에 참가하는 인원 m, 튜브의 순서 p
string solution(int n, int t, int m, int p) {
    string answer = "";
    
    string totalNum = ""; // 진법 변환한 수 다 합친 결과
    int current = 0;
    while(totalNum.length() < m*t){
        totalNum += change(current++, n);
    }
    
    // 튜브 결과 넣기
    for(int i = 0; i < t; i++){
        answer += totalNum[m*i+(p-1)];
    }
    
    return answer;
}

File: 프로그래머스/2/17687. ［3차］ n진수 게임/［3차］ n진수 게임.java
import java.util.*;

class Solution {
    
    private static final String DIGITS = "0123456789ABCDEF";
    
    // 10진수 from => n진수 string으로 변환
    public String toN(int n, int from){
        StringBuilder ans = new StringBuilder();
        if(from == 0) return "0"; 
        while(from > 0){
            ans.append(DIGITS.charAt(from % n)); 
            from /= n;
        }
        // 거꾸로
        return ans.reverse().toString();
    }
 
                                        
    // 진법, 구할 숫자 갯수, 참가 인원, 순서
    public String solution(int n, int t, int m, int p) {
        int cur = 0; 
        
        StringBuilder numbers = new StringBuilder(); 
        // String으로 모든 수 기록 -> t*m 길이가 될 때까지
        while(numbers.length() < t*m){
            // 10진수 -> N진법으로 변환
            numbers.append(toN(n, cur)); 
            cur++;
        }
        
        // 말해야 하는 숫자 구하기
        StringBuilder answer = new StringBuilder();
        for(int i = p-1; i < t*m; i+=m){
            answer.append(numbers.charAt(i));
        }

        // 말해야 하는 숫자 출력
        return answer.toString();
    }
}